Test suite v0.4

teocns opened the pull request:
## Overview
This PR reorganizes the test infrastructure by implementing a comprehensive testing strategy that separates unit and integration tests, with a focus on provider integrations and API functionality testing.

## Infrastructure Changes
- [x] Split tests into dedicated `unit/` and `integration/` directories
- [x] Configure separate CI jobs for unit and integration tests
- [x] Add VCR.py for HTTP interaction recording/replay
- [x] Centralize test fixtures in `conftest.py`
- [x] Add development tools (pytest-sugar, pdbpp)
- [x] Update pytest configuration for optimal test isolation
- [x] Set up proper test timeouts (5 minutes)

## Test Fixtures
- [x] Implement JWT token management
- [x] Add provider response spying capabilities
- [x] Centralize `mock_req` fixture
- [x] Add session management utilities
- [x] Add package availability control
- [x] Set up VCR ignore hosts and options

## Core Functionality Tests
- [x] Improve session handling and teardown
- [x] Enhance async loop lifecycle management
- [x] Add concurrent API request handling tests
- [x] Implement proper singleton cleanup
- [x] Move telemetry tests to unit test directory

## Provider Integration Tests
### OpenAI
- [x] Basic sync/async completions
- [x] Streaming responses
- [x] Assistants API integration

### Mistral
- [x] Sync completions
- [x] Async completions
- [x] Streaming responses

### Cohere
- [x] Chat completions (sync)
- [x] Chat completions (async)
- [x] Stream handling
- [x] Instrumentation control

### AI21
- [x] Sync completions
- [x] Async completions
- [x] Stream management

### Groq
- [x] Basic completions
- [x] Session management
- [x] Stream handling

### LlamaStack
- [ ] Agent configuration
- [ ] Shield management
- [ ] Model selection

## CrewAI Integration Tests
- [ ] Basic Setup
  - Test initialization
  - Test session creation
  - Test auto-end behavior
- [ ] Session Management
  - Test crew lifecycle
  - Test task completion tracking
  - Test manual session control
- [ ] Agent Monitoring
  - Test single agent tracking
  - Test multi-agent tracking
  - Test task delegation tracking
- [ ] Tool Integration
  - Test built-in tools
  - Test custom tools
  - Test tool error handling
- [ ] Example Workflows
  - Test job posting flow
  - Test markdown validator
  - Test Instagram post creation

## API Server Tests
- [x] Session lifecycle in API context
- [x] Tool recording
- [x] Event validation
- [x] Multi-session handling

## Documentation
- [ ] Update CONTRIBUTING.md with new test structure
- [ ] Document VCR usage and best practices
- [ ] Add provider test documentation

## Integration
- [x] Adapt and integrate tests from PR #603
- [ ] Update LangChain handler tests for current API
- [ ] Implement remaining provider integration tests

## CrewAI Integration Tests

### Basic Integration Tests
- [ ] Test initialization with CrewAI
  - Verify AgentOps initialization before Crew constructor
  - Test auto_start_session behavior
  - Test skip_auto_end_session parameter
  - Verify proper session creation

### Session Management Tests
- [ ] Test session lifecycle with CrewAI
  - Verify session starts with Crew initialization
  - Test automatic session ending when tasks complete
  - Verify session state after crew.kickoff()
  - Test manual session ending

### Event Recording Tests
- [ ] Test LLM event recording
  - Verify LLM calls are properly tracked
  - Test streaming responses tracking
  - Verify async LLM calls tracking
  - Test multiple LLM calls within one session

### Multi-Agent Tests
- [ ] Test multi-agent scenarios
  - Verify each agent's actions are tracked
  - Test inter-agent communication tracking
  - Verify task delegation and completion tracking
  - Test parallel agent execution monitoring

### Tool Usage Tests
- [ ] Test tool integration
  - Verify custom tool execution tracking
  - Test built-in tool usage monitoring
  - Verify tool error handling
  - Test tool result recording

### Example Workflow Tests
- [ ] Test job posting workflow
  - Verify researcher agent tracking
  - Test writer agent monitoring
  - Verify review agent tracking
  - Test complete workflow execution

### Error Handling Tests
- [ ] Test error scenarios
  - Verify failed task tracking
  - Test exception handling
  - Verify session state after errors
  - Test recovery mechanisms

### Integration with Other Tools
- [ ] Test CrewAI with other integrations
  - Test OpenAI provider integration
  - Verify LangChain compatibility
  - Test custom tool implementations
  - Verify third-party tool usage

### Special Features Tests
- [ ] Test CrewAI-specific features
  - Verify task dependency tracking
  - Test sequential vs parallel execution
  - Verify agent role assignments
  - Test custom agent configurations
  
  
## Known Issues & Mitigations
- [x] Test timing out when running integration and unit tests together
  - **Solution**: Separated test runs in CI
- [x] Complex patching layers management
  - **Solution**: Centralized mock configuration
- [x] VCR initialization conflicts
  - **Solution**: Scoped VCR config to session level

teocns committed:
test: add llm_event_spy fixture for tests

teocns committed:
test: add VCR.py fixture for HTTP interaction recording

teocns committed:
test: add `__init__` to make `tests/` a package

teocns committed:
deps: group integration-testing

teocns committed:
test: add fixture to mock package availability in tests

teocns committed:
test: Add integration tests for OpenAI provider and features

teocns committed:
test: add tests for concurrent API requests handling

teocns committed:
Improve vcr.py configuration

Signed-off-by: Teo <teocns@gmail.com>

teocns committed:
ruff

Signed-off-by: Teo <teocns@gmail.com>

teocns committed:
chore(pyproject): update pytest options and loop scope

teocns committed:
chore(tests): update vcr.py ignore_hosts and options

teocns commented:
Some funny behavior with what I believe is the async loop lifecycle management. Tests time out, or maybe there is one particular test making the testing suite time out. Running each individually work, so far. Tried different combos of async event loop configuration without luck

teocns commented:
I found out the issue is hit only when vcr is initialized (i.e imported)

Tests keep going until a vcr replay kicks in

I think this has to do with Async httpx

teocns committed:
pyproject.toml

Signed-off-by: Teo <teocns@gmail.com>

teocns committed:
centralize teardown in conftest.py (clear singletons, end all sessions)

Signed-off-by: Teo <teocns@gmail.com>

teocns committed:
change vcr_config scope to session

Signed-off-by: Teo <teocns@gmail.com>

teocns committed:
integration: auto start agentops session

Signed-off-by: Teo <teocns@gmail.com>

teocns committed:
Move unit tests to dedicated folder (tests/unit)

Signed-off-by: Teo <teocns@gmail.com>

teocns committed:
Isolate vcr_config import into tests/integration

Signed-off-by: Teo <teocns@gmail.com>

teocns committed:
configure pytest to run only unit tests by default, and include integration tests only when explicitly specified.

Signed-off-by: Teo <teocns@gmail.com>

teocns committed:
ci(python-tests): separate job between unit-integration tests

teocns committed:
set python-tests timeout to 5 minutes

Signed-off-by: Teo <teocns@gmail.com>

teocns committed:
ruff

Signed-off-by: Teo <teocns@gmail.com>

teocns committed:
Implement jwt fixture, centralized reusable mock_req into conftest.py

Signed-off-by: Teo <teocns@gmail.com>

reauthorize

Signed-off-by: Teo <teocns@gmail.com>

teocns committed:
ci(python-tests): simplify env management, remove cov from integration-teests

Signed-off-by: Teo <teocns@gmail.com>

teocns committed:
ruff

Signed-off-by: Teo <teocns@gmail.com>

teocns commented:
Yeap, a lot of things are going on here: provider patching (agentops), requests patching (vcr), requests-mock (pytest)... 

I found that running integration tests separately from unit tests works, hence we'll keep it that way for now.

teocns committed:
fix: cassette for test_concurrent_api_requests

Signed-off-by: Teo <teocns@gmail.com>

teocns commented:
We could use devin's tests now, just need a bit of work around #603 to adapt it, if anyone would like to take over. I'm moving onto merging everything on [feat/v0.4](https://github.com/AgentOps-AI/agentops/tree/feat/v0.4)

teocns commented:
CONTIRBUTING.md needs to be updated aswell

areibman commented on tests/fixtures/vcr.py:
```diff
@@ -0,0 +1,119 @@
+        #             str(Path.home()), "REDACTED_HOME_PATH"
+        #         )
+        #     } if isinstance(response.get("body", {}).get("string"), str) else response["body"]
+        # }
```
What's this for?

teocns replied:
IDK

teocns replied:
Never worked

teocns replied:
Speculative functionality. Wanted to filter out sensitive host env data but we can just edit the cassette directly

areibman commented:
All seems to work locally. Mostly just moving fixtures and replacing requests with VCR as far as I can see

teocns committed:
Cleanup vcr.py comments

Signed-off-by: Teo <teocns@gmail.com>

teocns commented:
This is ready for merge if @the-praxs has no additional comments on this

dot-agi commented:
> This is ready for merge if @the-praxs has no additional comments on this

Reviewing and going to get this merged, will fix issues if any along the way.

dot-agi committed:
add a `TODO` for removing `vcrpy` git version after its release

dot-agi reviewed (changes requested):
I have observed the following that needs changes:
1. LangChain handler test was already outdated since it's using old API references of `agentops` and `langchain`.
2. OpenAI Assistants are handled by a different method `handle_assistants_function` which is absent in the fixture and integration tests.
3. Integration tests for other LLM providers are missing.
4. Missing docs for this new testing suite.

I am working on these so that we can have an almost fully-tested suite.

dot-agi committed:
refactor openai assistants response handling for easier testing

dot-agi committed:
add more keys for different llm providers

dot-agi committed:
add integration tests for other providers

dot-agi committed:
remove openai version limitation

dot-agi committed:
add providers as deps

teocns committed:
chore: add mistralai to test dependencies

dot-agi committed:
remove `mistral` from dependency since its incorrect

dot-agi commented:
So far we have the integration tests for all providers except Llama Stack and the partners (crewAI, Autogen, TaskWeaver, LangChain).

I have Llama Stack configured with Fireworks to make it work on my local machine but that's unreliable. I need to make it work to get the cassette recordings.

Partner integration tests I am working on the comprehensive suite as mentioned above.

dot-agi committed:
ruff

dot-agi committed:
re-record cassettes

teocns committed:
tests/fixtures/providers: fallback to `test-api-key` if no provider is found

all provider fixtures will:
Use the actual API key if it's set in the environment
Fall back to "test-api-key" if no environment variable is found

Signed-off-by: Teo <teocns@gmail.com>

dot-agi committed:
set keys for `litellm`

teocns committed:
Improve tests/integration/test_llm_providers.py openai assistants

Signed-off-by: Teo <teocns@gmail.com>

teocns committed:
Make integration tests appropriately skip, regenerate x1 cassette

Signed-off-by: Teo <teocns@gmail.com>

teocns committed:
explicit tests/integration/conftest finxtures import

Signed-off-by: Teo <teocns@gmail.com>

teocns committed:
deps: improve dev packages versionings

teocns committed:
Make integration tests run with python 3.12

Signed-off-by: Teo <teocns@gmail.com>

teocns committed:
add uv.lock

Signed-off-by: Teo <teocns@gmail.com>

teocns committed:
test concurrent api requests: remove matcher on method, possibly causing contingent error

Signed-off-by: Teo <teocns@gmail.com>

teocns committed:
Run static-analysis with python 3.12.2

Signed-off-by: Teo <teocns@gmail.com>
